rest/oidc_api.go

```diff
@@ -118,6 +146,28 @@ func (h *handler) handleOIDCCallback() error {
 		return base.HTTPErrorf(http.StatusBadRequest, "Unable to identify provider for callback request")
 	}
 
+	// Validate state parameter to prevent cross-site request forgery (CSRF) when callback state is enabled.
+	if !provider.DisableCallbackState {
+		stateCookie, err := h.rq.Cookie(stateCookieName)
+
+		if err == http.ErrNoCookie || stateCookie == nil {
+			return ErrNoStateCookie
+		}
+
+		if err != nil {
+			base.Warnf("Unexpected error attempting to read OIDC state cookie: %v", err)
+			return ErrReadStateCookie
+		}
+
+		stateParam := h.rq.URL.Query().Get(requestParamState)
+		if stateParam != stateCookie.Value {
+			return ErrStateMismatch
+		}
+
+		// Delete the state cookie on successful validation.
+		h.deleteStateCookie(stateCookie)
```

Comment:
Performing a 'SetCookie' based on an incoming cookie feels like an unnecessary security risk (we're blindly returning the incoming value), and I'm pretty sure clients don't send all the cookie properties that were originally included in Set-Cookie.  Is there a reason to not just refactor setStateCookie so that it can be used here as well?

Reply:
Thank you Adam, removed deleteStateCookie and refactored setStateCookie in latest commit.